Clippy pointer cast lint experiment

Ryan1729 opened the pull request:
This PR is an experiment about exposing more parts of `rustc_typeck` for use in `clippy`. In particular, the code that checks where a cast is valid or not was exposed, which necessitated exposing [`FnCtxt`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_typeck/check/struct.FnCtxt.html), and figuring out how to create an instance of that type inside `clippy`.

This was prompted by [this clippy issue](https://github.com/rust-lang/rust-clippy/issues/2064).

r? @oli-obk

Ryan1729 committed:
make parts of rustc_typeck public

Ryan1729 committed:
run cargo dev new_lint then move transmutes_expressible_as_ptr_casts into transmute module

Ryan1729 committed:
make InheritedBuilder::enter public

Ryan1729 committed:
initial compiling version of TRANSMUTES_EXPRESSIBLE_AS_PTR_CASTS

Ryan1729 committed:
write currently failing test for transmutes_expressible_as_ptr_casts

There are 5 errors, when there should be 7.

Ryan1729 committed:
accidentally cause an ICE by putting the TRANSMUTES_EXPRESSIBLE_AS_PTR_CASTS handling after the match

The reason I did this in the first place was to try and figure out why I don't see my expected 7 error messages

Ryan1729 committed:
try putting the can_be_expressed_as_pointer_cast at the top and find that we still get an ICE

Ryan1729 committed:
get the expected number of errors by acknowledging that other lints are covering the same ground

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @oli-obk (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://rustc-dev-guide.rust-lang.org/contributing.html) for more information.

oli-obk commented on src/tools/clippy/clippy_lints/src/transmute.rs:
```diff
@@ -648,3 +693,57 @@ fn is_layout_incompatible<'tcx>(cx: &LateContext<'tcx>, from: Ty<'tcx>, to: Ty<'
+        let fn_ctxt = FnCtxt::new(
+            &inherited,
+            // TODO should we try to get the correct ParamEnv?
+            ty::ParamEnv::empty(),
```
yes, please use `cx.param_env`

Ryan1729 replied:
If I knew it was that easy, I would have done it from the start. I meant to get back to the param_env part after I figured out how to create an `Inherited`, but I forgot.

oli-obk commented on src/tools/clippy/clippy_lints/src/transmute.rs:
```diff
@@ -648,3 +693,57 @@ fn is_layout_incompatible<'tcx>(cx: &LateContext<'tcx>, from: Ty<'tcx>, to: Ty<'
+            None
+        }
+    })
+}
```
missing trailing newline

oli-obk commented on src/tools/clippy/clippy_lints/src/transmute.rs:
```diff
@@ -648,3 +693,57 @@ fn is_layout_incompatible<'tcx>(cx: &LateContext<'tcx>, from: Ty<'tcx>, to: Ty<'
+            hir_id
+        );
+
+        // If we already have errors, we can't be sure we can pointer cast.
```
can this happen? Just the creation shouldn't set this flag, right? Maybe just `assert` here so we know if/when it can happen

Ryan1729 replied:
I have not observed this happening. Asserting here sounds good to me.

oli-obk commented on src/tools/clippy/clippy_lints/src/transmute.rs:
```diff
@@ -648,3 +693,57 @@ fn is_layout_incompatible<'tcx>(cx: &LateContext<'tcx>, from: Ty<'tcx>, to: Ty<'
+                .ok()
+                // do_check's documentation says that it might return Ok and create
+                // errors in the fcx instead of returing Err in some cases.
+                .filter(|_| !fn_ctxt.errors_reported_since_creation())
```
This seems bad... won't we start emitting errors for weird transmutes?

Ryan1729 replied:
My intention was that if errors happened, we would return `None`. As far as I understand, that `filter` achieves that since I negated the boolean. So if error happen in the `FnCtxt`, this code should be treating it the same as an `Err` result, unless I've missed something.

I assumed that if any errors happened in the `FnCtxt`, some other check, probably from `rustc` itself, would catch the problem and report it. Does that seem like a reasonable assumption, and does that fit with your expectations?

oli-obk replied:
My worry is that the errors have already been emitted. I haven't checked how errors could be emitted, so... I guess look for these errors and try to use the arguments on which they'd error for a cast on a transmute and see how the tests fare.

Ryan1729 replied:
Of all the functions called from inside `do_check`, `check_ref_cast` is the one that most clearly has a case where errors are emitted. This is because it calls [`demand_eqtype`](https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_typeck/check/demand.rs.html#72-76) which clearly can emit an error.

The following functions clearly do not emit errors:
* `CheckCast::check_ptr_ptr_cast`
* `CheckCast::check_ptr_addr_cast`
* `CheckCast::check_addr_ptr_cast`
* `CheckCast::check_fptr_ptr_cast`

I haven't confirmed whether or not the other functions called in `do_check` emit errors though.

I also had a look at the [commit the introduced `do_check` and the doc comment on it I mentioned above](https://github.com/rust-lang/rust/commit/83acebc46213e30eac3e9e71b213440104914f4c), and `check_ref_cast` was already called within `do_check` at that time, so I suspect that is what the doc comment originally referred to.

So, my next step is to add code to the tests that triggers `demand_eqtype` to emit an error, and see what happens.

Ryan1729 replied:
I tried adding a `dbg!` call into `demand_eqtype` so that if the error case was called, the tests would definitely fail, (due to the extra `stdout` output.) That way, we would know that the new tests would actually hit that case, so we would know whether or not we got an extra error message.

The existing tests for the new lint did not hit that case, but from looking at the code leading from `do_check` to `demand_eqtype`, it seems like a `transmute` from an array reference to a pointer to a different type should trigger the lint to  call `do_check` in a way that should trigger it. For example, `&[i32; 1]` to `*const u8`. So I added such a `transmute` to the tests, but I got an error from a different transmute lint.

While the other lint seems to prevent any issues at the moment, if that lint, or the ordering was changed in the future that could lead to a surprising bug, so it seems prudent to understand whether the `do_check` call will cause an extra error or not.

So, I temporarily placed the new lint at the top of the big `match` inside the `check_expr` implementation inside `transmute.rs`. Doing that lead to an ICE with the following error message:

```
thread 'rustc' panicked at 'assertion failed: self.find(parent).map_or(false, |n| is_body_owner(n, hir_id))', src/librustc_middle/hir/map/mod.rs:352:9
```

So without digging into what the failed assert means exactly, I would guess that something is wrong with the way that the FnCtxt was created inside clippy in this PR. But I'm not really sure. Does anyone else have more context around what that assert failing suggests?

oli-obk replied:
> but I got an error from a different transmute lint.

If that other lint is `allow`ed, would the error occur? Or is it part of this big `match`, so we never reach your new lint if the other lint is emitted? I'd be fine to just leave sufficient comments (and add a test that will fail horribly if the order is changed).

Ryan1729 replied:
> is it part of this big match, so we never reach your new lint if the other lint is emitted?
That is the case.

Some other tests were already failing whenever I observed the ICE, but better safe than sorry. I've added a the transmute from `&[i32; 1]` to `*const u8` case to the lint-specific test.

oli-obk replied:
Ok great! I think you can now just make this an assert instead of filtering

Ryan1729 replied:
Sounds good. I added a not about the panics to the comments too. All that is in [this commit](https://github.com/rust-lang/rust/pull/75098/commits/54472478fa29c2d121da469bc8f10f8b68e3e134)

Ryan1729 committed:
address some review comments

Ryan1729 committed:
add description to assert

Ryan1729 committed:
add documentation to functions that call `do_check` and add a test against lint ordering changing

Ryan1729 committed:
add extra error message to the expected stderr for transmutes_expressible_as_ptr_casts test

oli-obk commented:
@rust-lang/clippy we're doing a clippy PR against rustc, because it also requires rustc changes. Now that the rustc changes work, we could also split it and move the clippy changes to a clippy PR, or we review it here and then sync afterwards. Opinions?

@Ryan1729 please rebase and ensure that none of your commits contain submodule changes

Ryan1729 committed:
change filter to assert, and update comments

oli-obk commented on src/tools/clippy/tests/ui/transmutes_expressible_as_ptr_casts.rs:
```diff
@@ -0,0 +1,76 @@
+// fall through into `do_check`.
+fn trigger_do_check_to_emit_error(in_param: &[i32; 1]) -> *const u8 {
+    unsafe { transmute::<&[i32; 1], *const u8>(in_param) }
+}
```
missing a newline at the end of the file

Ryan1729 committed:
add newline to transmutes_expressible_as_ptr_casts.rs

oli-obk commented:
You need to run `./x.py fmt` before CI even starts building the PR

Ryan1729 committed:
run ./x.py fmt

Ryan1729 committed:
run clippy_dev update_lints

Ryan1729 committed:
run clippy_dev fmt

This seemed to overdo it a bit, affecting multiple submodules, and changing a file I didn't touch, so I didn't commit those changes

flip1995 commented:
I can review the Clippy changes in this PR and we can sync afterwards. I don't think it is worth to go back and forth with syncs for changes that affect rustc and clippy (to avoid this we switched to `subtree` in the first place)

flip1995 commented on src/tools/clippy/clippy_lints/src/transmute.rs:
```diff
@@ -648,3 +693,59 @@ fn is_layout_incompatible<'tcx>(cx: &LateContext<'tcx>, from: Ty<'tcx>, to: Ty<'
     }
 }
+
+/// Check if the the type conversion can be expressed as a pointer cast, instead of
```
```suggestion
/// Check if the type conversion can be expressed as a pointer cast, instead of
```

flip1995 commented on src/tools/clippy/tests/ui/transmutes_expressible_as_ptr_casts.rs:
```diff
@@ -0,0 +1,76 @@
+#![warn(clippy::transmutes_expressible_as_ptr_casts)]
```
```suggestion
// run-rustfix
#![warn(clippy::transmutes_expressible_as_ptr_casts)]
```

I think the suggestion of this lint is auto applicable, so we should test it.

flip1995 commented on src/tools/clippy/clippy_lints/src/transmute.rs:
```diff
@@ -601,7 +628,25 @@ impl<'tcx> LateLintPass<'tcx> for Transmute {
+                        |diag| {
+                            if let Some(arg) = sugg::Sugg::hir_opt(cx, &args[0]) {
+                                let sugg = arg.as_ty(&to_ty.to_string()).to_string();
+                                diag.span_suggestion(e.span, "try", sugg, Applicability::Unspecified);
```
```suggestion
                                diag.span_suggestion(e.span, "try", sugg, Applicability::MachineApplicable);
```

I think this is `MachineApplicable`. The only reason everything in the `transmute` module is `Unspecified`, is that I was to lazy to check the `Applicability`, when this argument was introduced back then.

Or are there any concerns this suggestion is `MaybeIncorrect`?

Ryan1729 replied:
All the suggestions I have seen from this lint have been correct, as far as my understanding goes.

flip1995 reviewed (approved):
Clippy changes LGTM overall. Just a few NITs

Before merging this PR, I would like to test the changes in Clippy's CI, so we make sure, that it passes the integration tests of Clippy (looking for ICEs).

Ryan1729 committed:
Apply suggestions from code review

Co-authored-by: Philipp Krones <hello@philkrones.com>

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit d897fd28ae46b031a67814ffe29922b14bd028d4 has been approved by `oli-obk`

<!-- @bors r=oli-obk d897fd28ae46b031a67814ffe29922b14bd028d4 -->
<!-- homu: {"type":"Approved","sha":"d897fd28ae46b031a67814ffe29922b14bd028d4","approver":"oli-obk"} -->

bors commented:
:hourglass: Testing commit d897fd28ae46b031a67814ffe29922b14bd028d4 with merge 61793bf48b713076d2c7e35e22f7bf4fc539172c...
<!-- homu: {"type":"BuildStarted","head_sha":"d897fd28ae46b031a67814ffe29922b14bd028d4","merge_sha":"61793bf48b713076d2c7e35e22f7bf4fc539172c"} -->

flip1995 commented:
@oli-obk This will fail the clippy UI-tests. @Ryan1729 you have to produce a `*.fixed` reference file, like the `*.stderr` reference file, so that the UI test passes. (`sh tests/ui/update-all-references.sh`) and then commit the `*.fixed` file.

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/954860536)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/954860536","builder_name":"checks-actions"} -->

Manishearth commented:
I am also in favor of not splitting it. We should be comfortable making substantative clippy changes from this repo when we feel it appropriate.

Ryan1729 committed:
copy over *.fixed file

Ryan1729 commented:
@flip1995 If I run the clippy tests by running `./x.py test clippy` then the `*.fixed` files end up in `./build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/build/clippy-<hash>/out/test_build_base/` so that script doesn't find them. So I had to copy the file manually.

oli-obk commented:
@bors r+

we should probably make `./x.py test src/tools/clippy --bless` work $somehow

bors commented:
:pushpin: Commit c04c4cb6e4b02b44501ab8e136616cbccaabd602 has been approved by `oli-obk`

<!-- @bors r=oli-obk c04c4cb6e4b02b44501ab8e136616cbccaabd602 -->
<!-- homu: {"type":"Approved","sha":"c04c4cb6e4b02b44501ab8e136616cbccaabd602","approver":"oli-obk"} -->

flip1995 commented:
> we should probably make ./x.py test src/tools/clippy --bless work $somehow

Related to this: rust-lang/rust-clippy#5394

bors commented:
:hourglass: Testing commit c04c4cb6e4b02b44501ab8e136616cbccaabd602 with merge 4e1f1d3a7ea7f2352eb8b975c74b12fd0ca02b3f...
<!-- homu: {"type":"BuildStarted","head_sha":"c04c4cb6e4b02b44501ab8e136616cbccaabd602","merge_sha":"4e1f1d3a7ea7f2352eb8b975c74b12fd0ca02b3f"} -->

rust-log-analyzer commented:
The job `i686-gnu-nopt` of your PR [failed](https://github.com/rust-lang-ci/rust/runs/959359201) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://github.com/rust-lang-ci/rust/commit/4e1f1d3a7ea7f2352eb8b975c74b12fd0ca02b3f/checks/959359201/logs), [raw log](https://github.com/rust-lang-ci/rust/commit/4e1f1d3a7ea7f2352eb8b975c74b12fd0ca02b3f/checks/959359201/logs)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain

```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@rust-lang/infra`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/959376965)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/959376965","builder_name":"checks-actions"} -->

Ryan1729 commented:
Here’s a snippet from the bottom of the raw log linked above that suggests the failure was caused by a network issue. 

```
2020-08-07T17:55:39.4818196Z Caused by:
2020-08-07T17:55:39.4818413Z   failed to get 200 response from `https://crates.io/api/v1/crates/maplit/1.0.1/download`, got 502
```

oli-obk commented:
@bors retry

bors commented:
:hourglass: Testing commit c04c4cb6e4b02b44501ab8e136616cbccaabd602 with merge d72af5f4695cbd0c3d565eb53c4cdb50a5370e6b...
<!-- homu: {"type":"BuildStarted","head_sha":"c04c4cb6e4b02b44501ab8e136616cbccaabd602","merge_sha":"d72af5f4695cbd0c3d565eb53c4cdb50a5370e6b"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/961876990)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/961876990","builder_name":"checks-actions"} -->

oli-obk commented:
clippy's stderr files need updating again, you may need to rebase first

Ryan1729 committed:
update stderr for transmutes_expressible_as_ptr_casts

Ryan1729 committed:
add a test example of where transmutes_expressible_as_ptr_casts should not suggest anything

Ryan1729 committed:
fix unary minus on usize and unused variable errors in .fixed file

Ryan1729 committed:
add allow unused_unsafe and allow dead_code

Ryan1729 commented:
After several small changes, this is passing again, locally.

Two changes are most notable:
The fixed code wasn't compiling because I was getting a warning about `use std::mem::transmute`` being unused after the fixes removed all calls to it. I noticed then that a situation where no suggestion should happen was not part of the tests. So I added a new test that containing a `transmute` call that should not and does not get removed by the fixer.

The other notable change was that I changed the use of `-1isize` in the new tests to `usize::MAX` since otherwise the fixed version resulted in `-1 as *const i32` which does not compile since the `1` ia inferred to be a `usize`. I couldn't find an easy way to change the suggestion to include the type suffix, so since actually wanting to use a unary minus to write a number to be cast as a pointer seems rather niche, (I only used `-1` here so if something went wrong it would be obvious), I just changed the type to be a `usize` and kept the same value.

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit d2e7293078076c801a3774cdb0c805da3e1ba154 has been approved by `oli-obk`

<!-- @bors r=oli-obk d2e7293078076c801a3774cdb0c805da3e1ba154 -->
<!-- homu: {"type":"Approved","sha":"d2e7293078076c801a3774cdb0c805da3e1ba154","approver":"oli-obk"} -->
